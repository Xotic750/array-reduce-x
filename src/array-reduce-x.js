import attempt from 'attempt-x';
import splitIfBoxedBug from 'split-if-boxed-bug-x';
import toLength from 'to-length-x';
import toObject from 'to-object-x';
import assertIsFunction from 'assert-is-function-x';
import toBoolean from 'to-boolean-x';
import requireObjectCoercible from 'require-object-coercible-x';

const natRed = [].reduce;
const castObject = {}.constructor;
const nativeReduce = typeof natRed === 'function' && natRed;

const test1 = function test1() {
  return attempt.call([], nativeReduce, function attemptee(acc) {
    return acc;
  }).threw;
};

const test2 = function test2() {
  const res = attempt.call(
    castObject('abc'),
    nativeReduce,
    function attemptee(acc, c) {
      return acc + c;
    },
    'x',
  );

  return res.threw === false && res.value === 'xabc';
};

const test3 = function test3() {
  const res = attempt.call(
    (function getArgs() {
      /* eslint-disable-next-line prefer-rest-params */
      return arguments;
    })(1, 2, 3),
    nativeReduce,
    function attempte(acc, arg) {
      return acc + arg;
    },
    1,
  );

  return res.threw === false && res.value === 7;
};

const test4 = function test4() {
  const res = attempt.call(
    {0: 1, 1: 2, 3: 3, 4: 4, length: 4},
    nativeReduce,
    function attempte(acc, arg) {
      return acc + arg;
    },
    2,
  );

  return res.threw === false && res.value === 8;
};

const test5 = function test5() {
  const doc = typeof document !== 'undefined' && document;

  if (doc) {
    const fragment = doc.createDocumentFragment();
    const div = doc.createElement('div');
    fragment.appendChild(div);

    const atemptee = function attempte(acc, node) {
      acc[acc.length] = node;

      return acc;
    };

    const res = attempt.call(fragment.childNodes, nativeReduce, atemptee, []);

    return res.threw === false && res.value.length === 1 && res.value[0] === div;
  }

  return true;
};

const test6 = function test6() {
  const res = attempt.call('ab', nativeReduce, function attempte() {
    /* eslint-disable-next-line prefer-rest-params */
    return arguments[3];
  });

  return res.threw === false && typeof res.value === 'object';
};

// ES5 15.4.4.21
// http://es5.github.com/#x15.4.4.21
// https://developer.mozilla.org/en/Core_JavaScript_1.5_Reference/Objects/Array/reduce
const isWorking = toBoolean(nativeReduce) && test1() && test2() && test3() && test4() && test5() && test6();

const patchedReduce = function reduce(array, callBack /* , initialValue */) {
  requireObjectCoercible(array);
  const args = [assertIsFunction(callBack)];

  if (arguments.length > 2) {
    /* eslint-disable-next-line prefer-rest-params,prefer-destructuring */
    args[1] = arguments[2];
  }

  return nativeReduce.apply(array, args);
};

const implementation = function reduce(array, callBack /* , initialValue */) {
  const object = toObject(array);
  // If no callback function or if callback is not a callable function
  assertIsFunction(callBack);
  const iterable = splitIfBoxedBug(object);
  const length = toLength(iterable.length);
  const argsLength = arguments.length;

  // no value to return if no initial value and an empty array
  if (length === 0 && argsLength < 3) {
    throw new TypeError('Reduce of empty array with no initial value');
  }

  let i = 0;
  let result;

  if (argsLength > 2) {
    /* eslint-disable-next-line prefer-rest-params,prefer-destructuring */
    result = arguments[2];
  } else {
    do {
      if (i in iterable) {
        result = iterable[i];
        i += 1;
        break;
      }

      // if array contains no values, no initial value to return
      i += 1;

      if (i >= length) {
        throw new TypeError('Reduce of empty array with no initial value');
      }
    } while (true); /* eslint-disable-line no-constant-condition */
  }

  while (i < length) {
    if (i in iterable) {
      result = callBack(result, iterable[i], i, object);
    }

    i += 1;
  }

  return result;
};

/*
 * This method applies a function against an accumulator and each element in the
 * array (from left to right) to reduce it to a single value.
 *
 * @param {Array} array - The array to iterate over.
 * @param {Function} callBack - Function to execute for each element.
 * @param {*} [initialValue] - Value to use as the first argument to the first
 *  call of the callback. If no initial value is supplied, the first element in
 *  the array will be used. Calling reduce on an empty array without an initial
 *  value is an error.
 * @throws {TypeError} If array is null or undefined.
 * @throws {TypeError} If callBack is not a function.
 * @throws {TypeError} If called on an empty array without an initial value.
 * @returns {*} The value that results from the reduction.
 */
const $reduce = isWorking ? patchedReduce : implementation;

export default $reduce;
