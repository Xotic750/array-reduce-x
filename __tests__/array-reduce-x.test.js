import $reduce, {implementation} from '../src/array-reduce-x';

const createArrayLike = function(arr) {
  const o = {};
  arr.forEach(function(e, i) {
    o[i] = e;
  });

  o.length = arr.length;

  return o;
};

[implementation, $reduce].forEach((reduce, testNum) => {
  describe(`reduce ${testNum}`, function() {
    let testSubject;

    beforeEach(function() {
      testSubject = [1, 2, 3];
    });

    it('is a function', function() {
      expect.assertions(1);
      expect(typeof reduce).toBe('function');
    });

    it('should throw when array is null or undefined', function() {
      expect.assertions(3);
      expect(function() {
        reduce();
      }).toThrowErrorMatchingSnapshot();

      expect(function() {
        reduce(void 0);
      }).toThrowErrorMatchingSnapshot();

      expect(function() {
        reduce(null);
      }).toThrowErrorMatchingSnapshot();
    });

    describe('array', function() {
      it('should pass the correct arguments to the callback', function() {
        expect.assertions(1);
        const spy = jest.fn().mockReturnValue(0);
        reduce(testSubject, spy);
        expect(spy).toHaveBeenCalledWith(1, 2, 1, testSubject);
      });

      it('should start with the right initialValue', function() {
        expect.assertions(1);
        const spy = jest.fn().mockReturnValue(0);
        reduce(testSubject, spy, 0);
        expect(spy).toHaveBeenCalledWith(0, 1, 0, testSubject);
      });

      it('should not affect elements added to the array after it has begun', function() {
        expect.assertions(2);
        const arr = [1, 2, 3];

        let i = 0;
        reduce(arr, function(a, b) {
          i += 1;

          if (i <= 4) {
            arr.push(a + 3);
          }

          return b;
        });

        expect(arr).toStrictEqual([1, 2, 3, 4, 5]);

        expect(i).toBe(2);
      });

      it('should work as expected for empty arrays', function() {
        expect.assertions(2);
        const spy = jest.fn();
        expect(function() {
          reduce([], spy);
        }).toThrowErrorMatchingSnapshot();

        expect(spy).not.toHaveBeenCalled();
      });

      it('should throw correctly if no callback is given', function() {
        expect.assertions(1);
        expect(function() {
          reduce(testSubject);
        }).toThrowErrorMatchingSnapshot();
      });

      it('should return the expected result', function() {
        expect.assertions(1);
        expect(
          reduce(testSubject, function(a, b) {
            return String(a || '') + String(b || '');
          }),
        ).toBe(testSubject.join(''));
      });

      it('should not directly affect the passed array', function() {
        expect.assertions(1);
        const copy = testSubject.slice();
        reduce(testSubject, function(a, b) {
          return a + b;
        });

        expect(testSubject).toStrictEqual(copy);
      });

      it('should skip non-set values', function() {
        expect.assertions(1);
        delete testSubject[1];
        const visited = {};
        reduce(testSubject, function(a, b) {
          if (a) {
            visited[a] = true;
          }

          if (b) {
            visited[b] = true;
          }

          return 0;
        });

        expect(visited).toStrictEqual({1: true, 3: true});
      });
    });

    describe('array-like objects', function() {
      let testObject;

      beforeEach(function() {
        testObject = createArrayLike(testSubject);
      });

      it('should pass the correct arguments to the callback', function() {
        expect.assertions(1);
        const spy = jest.fn().mockReturnValue(0);
        reduce(testObject, spy);
        expect(spy).toHaveBeenCalledWith(1, 2, 1, testObject);
      });

      it('should start with the right initialValue', function() {
        expect.assertions(1);
        const spy = jest.fn().mockReturnValue(0);
        reduce(testObject, spy, 0);
        expect(spy).toHaveBeenCalledWith(0, 1, 0, testObject);
      });

      it('should not affect elements added to the array after it has begun', function() {
        expect.assertions(2);
        const arr = createArrayLike([1, 2, 3]);

        let i = 0;
        reduce(arr, function(a, b) {
          i += 1;

          if (i <= 4) {
            arr[i + 2] = a + 3;
          }

          return b;
        });

        expect(arr).toStrictEqual({
          0: 1,
          1: 2,
          2: 3,
          3: 4,
          4: 5,
          length: 3,
        });

        expect(i).toBe(2);
      });

      it('should work as expected for empty arrays', function() {
        expect.assertions(2);
        const spy = jest.fn();
        expect(function() {
          reduce({length: 0}, spy);
        }).toThrowErrorMatchingSnapshot();

        expect(spy).not.toHaveBeenCalled();
      });

      it('should throw correctly if no callback is given', function() {
        expect.assertions(1);
        expect(function() {
          reduce(testObject);
        }).toThrowErrorMatchingSnapshot();
      });

      it('should return the expected result', function() {
        expect.assertions(1);
        expect(
          reduce(testObject, function(a, b) {
            return String(a || '') + String(b || '');
          }),
        ).toBe('123');
      });

      it('should not directly affect the passed array', function() {
        expect.assertions(1);
        const copy = createArrayLike(testSubject);
        reduce(testObject, function(a, b) {
          return a + b;
        });

        expect(testObject).toStrictEqual(copy);
      });

      it('should skip non-set values', function() {
        expect.assertions(1);
        delete testObject[1];
        const visited = {};
        reduce(testObject, function(a, b) {
          if (a) {
            visited[a] = true;
          }

          if (b) {
            visited[b] = true;
          }

          return 0;
        });

        expect(visited).toStrictEqual({1: true, 3: true});
      });
    });

    it('should have a boxed object as list argument of callback', function() {
      expect.assertions(2);

      let actual = void 0;
      reduce('foo', function(accumulator, item, index, list) {
        actual = list;
      });

      expect(typeof actual).toBe('object');
      expect(Object.prototype.toString.call(actual)).toBe('[object String]');
    });
  });
});
