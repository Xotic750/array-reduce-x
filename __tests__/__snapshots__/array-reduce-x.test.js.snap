// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reduce array should throw correctly if no callback is given 1`] = `"undefined is not a function"`;

exports[`reduce array should work as expected for empty arrays 1`] = `"Reduce of empty array with no initial value"`;

exports[`reduce array-like objects should throw correctly if no callback is given 1`] = `"undefined is not a function"`;

exports[`reduce array-like objects should work as expected for empty arrays 1`] = `"Reduce of empty array with no initial value"`;

exports[`reduce should throw when array is null or undefined 1`] = `"Array.prototype.reduce called on null or undefined"`;

exports[`reduce should throw when array is null or undefined 2`] = `"Array.prototype.reduce called on null or undefined"`;

exports[`reduce should throw when array is null or undefined 3`] = `"Array.prototype.reduce called on null or undefined"`;
